unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Animator/Handler/AnimatorComponentHandler.cs

```diff
@@ -0,0 +1,146 @@
+                    dataStore.animatorShapesReady.OnAdded += AnimatorShapeReadyAdded;
+                }
+            }
+        }
+
+        private void AnimatorShapeReadyAdded(long entityId, GameObject gameObject)
+        {
+            if (entityId != entity.entityId)
+                return;
+          
+            dataStore.animatorShapesReady.OnAdded -= AnimatorShapeReadyAdded;
+            
+            Initialize(entity);
+            isShapeLoaded = true;
+
+            UpdateAnimationState(model);
+        }
+
+        internal void Initialize(IDCLEntity entity)
+        {
+            if (entity == null)
+                return;
+
+            //NOTE(Brian): fetch all the AnimationClips in Animation component.
+            animComponent = entity.gameObject.transform.parent.GetComponentInChildren<Animation>(true);
+
+            if (animComponent == null)
+                return;
+
+            clipNameToClip.Clear();
+            int layerIndex = 0;
+
+            animComponent.playAutomatically = true;
+            animComponent.enabled = true;
+            animComponent.Stop(); //NOTE(Brian): When the GLTF is created by GLTFSceneImporter a frame may be elapsed,
+            //putting the component in play state if playAutomatically was true at that point.
+            animComponent.clip?.SampleAnimation(animComponent.gameObject, 0);
+
+            foreach (AnimationState unityState in animComponent)
+            {
+                clipNameToClip[unityState.clip.name] = unityState.clip;
+
+                unityState.clip.wrapMode = WrapMode.Loop;
+                unityState.layer = layerIndex;
+                unityState.blendMode = AnimationBlendMode.Blend;
+                layerIndex++;
+            }
+        }
+        
+        internal void UpdateAnimationState(PBAnimator model)
+        {
+            if (clipNameToClip.Count == 0 || animComponent == null)
+                return;
+
+            if (model.States.Count == 0)
+                return;
+            
+            for (int i = 0; i < model.States.Count; i++)
+            {
+                if (clipNameToClip.ContainsKey(model.States[i].Clip))
+                {
+                    AnimationState unityState = animComponent[model.States[i].Clip];
+                    unityState.weight = model.States[i].Weight;
+                    unityState.wrapMode = model.States[i].Loop ? WrapMode.Loop : WrapMode.Default;
+                    unityState.clip.wrapMode = unityState.wrapMode;
+                    unityState.speed = model.States[i].Speed;
+                    unityState.enabled = model.States[i].Playing;
+
+                    if (model.States[i].ShouldReset)
+                        ResetAnimation(model.States[i], unityState.clip);
+
+                    if (model.States[i].Playing && !animComponent.IsPlaying(model.States[i].Clip))
+                        animComponent.Play(model.States[i].Clip);
+                }
+            }
```

Comment:
maybe we could use this opportunity to do a proper refactor on this component? It was one of Brian's dreams.
cause I think this is the worst performing component we have. the code with, all those dictionaries that look almost redundant, is hard to follow and there are a lot of code for hotfixes there

Reply:
Sure! If you think that there is a more efficient approach go ahead! I will add this to the issues card we have for that kind of optimization.
https://github.com/decentraland/sdk/issues/341

I will merge this since it is a working component right now and since we didn't release the new ECS7 yet, we can change this without problems